runtime/stdlib/account.go

```diff
@@ -1585,24 +1586,48 @@ func changeAccountContracts(
 		oldCode, err := handler.GetAccountContractCode(location)
 		handleContractUpdateError(err)
 
+		memoryGauge := invocation.Interpreter.SharedState.Config.MemoryGauge
+		legacyUpgradeEnabled := invocation.Interpreter.SharedState.Config.LegacyContractUpgradeEnabled
+
 		oldProgram, err := parser.ParseProgram(
-			invocation.Interpreter.SharedState.Config.MemoryGauge,
+			memoryGauge,
 			oldCode,
 			parser.Config{
 				IgnoreLeadingIdentifierEnabled: true,
 			},
 		)
 
+		var legacyContractUpgrade bool
+		// if we are allowing legacy contract upgrades, fall back to the old parser when the new one fails
+		if !ignoreUpdatedProgramParserError(err) && legacyUpgradeEnabled {
```

Comment:
why do we need to parse with the new parser first and then fallback?
Can't we pick the correct parser straight-away based on  the `legacyUpgradeEnabled` flag?

Reply:
I don't think it's necessarily the case that people are going to want exactly one or the other in any given emulator instance; e.g. someone may want to test sequencing updates from v0.42 -> v1.0 with no business logic changes, followed immediately by a second update that applies the business logic changes we told them not to include. IMO this adds more flexibility that enables that use case.

Reply:
makes sense!